### PR TITLE
Bossrush updates

### PR DESCRIPTION
**Pull request by ScalarVector1**
This PR introduces some architecture for storing persistent data, which is utilized to do the following improvements on boss rush:

Requires all bosses to be beaten at least once before unlocking
high scores are saved between game sessions

In addition, the following changes have been made to boss rush:

Improved GUI with proper sounds, back buttons, and unlock progression tracker
Improved adaptive DR formula which prevents stall strategies

### CODE REVIEW

File: Core/Systems/BossRushSystem/BossRushGUIHack.cs [hunk 1]
```diff
@@ -26,13 +26,21 @@ private void UpdateBossMenu(On_Main.orig_Update orig, Main self, GameTime gameTi
 				if (inMenu)
 				{
 					inMenu = false;
-					Main.menuMode = 0;
+
+					Main.OpenWorldSelectUI();
+					Main.menuMode = 888;
```
Comment (TripleFate) [inline Core/Systems/BossRushSystem/BossRushGUIHack.cs:31]: replace instances of 888 with MenuID.FancyUI for readability

File: Core/Systems/PersistentDataSystem/PersistentDataStore.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+			string dir = Path.Join(Main.SavePath, "StarlightRiverPersistent");
+
+			if (!Directory.Exists(dir))
+				Directory.CreateDirectory(dir);
+
+			string currentPath = Path.Join(Main.SavePath, "StarlightRiverPersistent", GetType().Name);
+			LoadFromFile(currentPath);
+		}
+
+		public void Unload()
```
Comment (TripleFate) [inline Core/Systems/PersistentDataSystem/PersistentDataStore.cs:24]: remove empty method

**Review by TripleFate — CHANGES_REQUESTED** [review]
looks good but you should do a global search and replace of 
"= 888" and replace it with "= MenuID.FancyUI"

File: Core/Systems/PersistentDataSystem/PersistentDataStore.cs (hunk 2 above)
Comment (ScalarVector1) [reply Core/Systems/PersistentDataSystem/PersistentDataStore.cs:24]: This is required for interface conformance

**Review by TripleFate — APPROVED** [review]

**Pull request merged**